[Cai Zhi Bin] iP

czhi-bin opened the pull request:
# DukePro
> "It's not knowing what to do; it's doing what you know." - Tony Robbins [(source)](https://dansilvestre.com/productivity-quotes/)

DukePro frees your mind of having to remember things you need to do. It's
- text-based
- easy to learn
- *~FAST~ SUPER FAST* to use

All you need to do is,
1. download it from [here](https://nus-cs2103-ay2122s1.github.io/website/schedule/week4/project.html).
2. double-click it.
3. add your tasks.
4. let it manage your task for you 🤪 

And it is __FREE__!

Features:
- [X] Managing tasks
- [X] Managing deadlines
- [x] Managing events
- [X] Scolds the user
- [ ] Reminder (coming soon)

If you are a Java programmer, you can use it to practice java too. Here's the `main` method:
```java
public class Main {
    public static void main(String[] args) {
        Application.launch(MainApp.class, args);
    }
}
```

Commit:
Add Gradle support

czhi-bin committed:
Add greeting, echos, exit functions

czhi-bin committed:
Add enum Commands

czhi-bin committed:
Change BYE in enum Commands to EXIT to fit better

czhi-bin committed:
Add tasks attribute to store a list of tasks

czhi-bin committed:
Combine greet and echo into one start function

czhi-bin committed:
Change echo to getCommand to fit its function better

czhi-bin committed:
Add addTask and listTask function

czhi-bin committed:
Add Task class

czhi-bin committed:
Edit Duke class accordingly to fit Task class in

czhi-bin committed:
Add new getCommandType function and rename getCommand to processCommand

czhi-bin committed:
Add done functionality and integrate the enum into the functions

czhi-bin committed:
Change Task class to abstract and add inner class ToDo, Deadline, Event classes

czhi-bin committed:
Add TODO, DEADLINE, EVENT to the enum CommandTypes and edit functions accordingly

czhi-bin committed:
Change Event, ToDo Deadline from inner classes to standalone classes

czhi-bin committed:
Fix minor bug for the missing bracket

czhi-bin committed:
Add switch case for Event, Todo, Deadline and fix minor bugs

czhi-bin committed:
Add new DukeException class

czhi-bin committed:
Add 1 new type of DukeException

czhi-bin committed:
Add 1 new type of DukeException, total 2 types

czhi-bin committed:
Add ways to handle UnknownCommandException

czhi-bin committed:
Add functionality to handle ToDo task without description

czhi-bin committed:
Add support to delete task from the list

czhi-bin committed:
Add automated text ui testing

czhi-bin committed:
Change switch indentation to follow code standard

czhi-bin committed:
Add Save function

czhi-bin committed:
Add functionality to read date and time

czhi-bin committed:
Merge branch 'branch-Level-7'

czhi-bin committed:
Deconflict between level-7 and level-8

czhi-bin committed:
Remove unused method and fix some minor issues

czhi-bin committed:
Merge pull request #1 from czhi-bin/branch-Level-7

Remove unused method and fix some minor issues

czhi-bin committed:
Add new exception for invalid DateTime format

czhi-bin committed:
Update text-ui-test using LocalDateTime

czhi-bin committed:
Merge branch 'branch-Level-7' with new updates

czhi-bin committed:
Change TaskList to use LocalDateTime

czhi-bin committed:
Refactor output related function into Ui

czhi-bin committed:
Refactor task related function into TaskList

czhi-bin committed:
Refactor code into Parser to handle user input

czhi-bin committed:
Refactor codes into package

czhi-bin committed:
Add JUnit tests

czhi-bin committed:
Package the App as a JAR file

czhi-bin committed:
Add JavaDoc for Storage and Ui

czhi-bin committed:
Add JavaDoc for Parser

czhi-bin committed:
Add JavaDoc for the exceptions and TaskList

czhi-bin committed:
Add JavaDoc for different type of Task

czhi-bin committed:
Merge branch 'branch-A-JavaDoc'

czhi-bin committed:
Edit code to follow coding standard

czhi-bin committed:
Merge branch 'branch-A-CodingStandard'

czhi-bin committed:
Add function to find task by searching a keyword

czhi-bin committed:
Merge branch 'branch-Level-9'

czhi-bin committed:
Repackage task related class into one package

czhi-bin committed:
Add JavaDoc for Duke

czhi-bin committed:
Update JavaDoc for all class to new version

czhi-bin committed:
Refactor code into DukeCommand related class

czhi-bin committed:
Add NoSuchTaskException

czhi-bin committed:
Change code to use InvalidDateTimeException

czhi-bin committed:
Rename data package to exception which fits better

czhi-bin committed:
Refactor code related to creation of task

czhi-bin committed:
Add NoTaskDescriptionException and its handling

czhi-bin committed:
Change code according to coding standard

czhi-bin committed:
Add NoDateTimeException and its handling

czhi-bin committed:
Add NoCommandDescriptionException and its handling

czhi-bin committed:
Fix minor bug

huisthat commented on src/main/java/duke/parser/Parser.java:
```diff
@@ -0,0 +1,145 @@
+package duke.parser;
+
+import duke.command.*;
+import duke.exception.*;
+import duke.storage.Storage;
+import duke.task.*;
```
it is best to not use the * when importing, what do you think?

czhi-bin replied:
oh i think IntelliJ refractor code function auto combined all my imports into * again... thanks for the catch!

huisthat commented on src/main/java/duke/parser/Parser.java:
```diff
@@ -0,0 +1,145 @@
+     * @throws DukeException If some error occurred in the processing of user input,
+     *                       like the user inputted some command that duke does not support.
+     */
+    public DukeCommand processInput(String userInput) throws DukeException {
```
is it possible to have different methods to process the different inputs? might make it look less messy

czhi-bin replied:
hmmm you mean like one method for each case? But the methods will be like one-liner though

muhammad-faruq replied:
I think if you want to make it more convenient to receive multiple arguments you can use varargs. for the process input method

huisthat reviewed (commented):
good! just some suggestions about the Parser class

czhi-bin committed:
Change the import statements to single class import

czhi-bin committed:
Merge remote-tracking branch 'origin/add-gradle-support' into branch-A-Gradle

muhammad-faruq commented on src/main/java/duke/command/Add.java:
```diff
@@ -0,0 +1,40 @@
+ * @author Zhi Bin
+ * @version Duke Level 9
+ */
+public class Add extends DukeCommand {
```
To show that this class extends DukeCommand perhaps name it something like AddCommand to make it clear that this is a command.

muhammad-faruq commented on src/main/java/duke/command/Delete.java:
```diff
@@ -0,0 +1,43 @@
+ * @author Zhi Bin
+ * @version Duke Level 9
+ */
+public class Delete extends DukeCommand {
```
Similarly DeleteCommand

muhammad-faruq commented on src/main/java/duke/command/Done.java:
```diff
@@ -0,0 +1,41 @@
+ * @author Zhi Bin
+ * @version Duke Level 9
+ */
+public class Done extends DukeCommand {
```
Similarly DoneCommand

muhammad-faruq commented on src/main/java/duke/command/List.java:
```diff
@@ -0,0 +1,34 @@
+ * @author Zhi Bin
+ * @version Duke Level 9
+ */
+public class List extends DukeCommand {
```
Similarly ListCommand. This is where it is quite clear that this may be confusing. When creating this object we will create List list = new List() which is not very clear that this is in fact a Command rather than a list datastructure

muhammad-faruq reviewed (commented):
Not much coding standard violation

muhammad-faruq commented on src/main/java/duke/exception/DukeException.java:
```diff
@@ -0,0 +1,28 @@
+     * @return The formatted error message.
+     */
+    protected String formatMessage(String msg) {
+        String HORIZONTAL_LINE = "____________________________________________________________";
```
If everything has indentation, why is indentation a separate thing? Setting indentation separately assumes that some lines do not need indentation but this does not seem to be the case.

muhammad-faruq commented on src/main/java/duke/parser/Parser.java:
```diff
@@ -0,0 +1,160 @@
+        int spaceIndex = userInput.indexOf(" ");
+        String taskDescription = userInput.substring(spaceIndex + 1);
+        if (taskDescription.isBlank() || spaceIndex == -1) {
+            throw new NoTaskDescriptionException();
```
Its a bit odd that the processing of task descriptions happen here. While the processing of additional information for other commands happen elsewhere. Perhaps make it consistent and make the processing of the additional information all happen in the commands while the parser only checks for the type of command

muhammad-faruq commented on src/main/java/duke/task/Deadline.java:
```diff
@@ -0,0 +1,59 @@
+     *
+     * @return The string representation of the deadline task.
+     */
+    @Override
```
Does this toString method give the deadline in the proper format? I am not too sure

muhammad-faruq commented on src/main/java/duke/task/Task.java:
```diff
@@ -0,0 +1,69 @@
+     * @param s The keyword to search for in the description.
+     * @return True if the description contains the keyword, else false.
+     */
+    public boolean findKeyword(String s) {
```
Perhaps here instead of just finding the keyword, we can make it more usable such that it will find the keyword regardless of capitalization. So things such as "find Work" will also return tasks that contains "work" or "wOrK".

muhammad-faruq commented on src/main/java/duke/ui/Ui.java:
```diff
@@ -0,0 +1,120 @@
+     * Introduction message with duke's logo.
+     */
+    private static final String INTRO = "Hello from\n"
+            + INDENTATION + " ____        _        \n"
```
Perhaps personalize this DUKE logo ><

muhammad-faruq reviewed (commented):
Code quality review.

czhi-bin committed:
Add Gradle to automate project build

czhi-bin committed:
Merge branch 'branch-A-Gradle'

czhi-bin committed:
Add checkstyle for Gradle

czhi-bin committed:
Merge branch 'branch-A-CheckStyle'

czhi-bin committed:
Add GUI with only echoing function

czhi-bin committed:
Update UI and command code for GUI

czhi-bin committed:
Rename command class name for more clarity

czhi-bin committed:
Customise Duke GUI

czhi-bin committed:
Remove formatMessage function in exceptions

czhi-bin committed:
Add auto exit after user typed bye

czhi-bin committed:
Merge branch 'branch-Level-10'

czhi-bin committed:
Fix checkstyle issues

czhi-bin committed:
Add Assertions in Duke and MainWindow

czhi-bin committed:
Improve code quality for Parser

czhi-bin committed:
Fix bug where invalid task number will cause duke to crash

czhi-bin committed:
Fix bug where program does not terminate after bye

czhi-bin committed:
Fix checkstyle issues

czhi-bin committed:
Add usage of streams in TaskList

czhi-bin committed:
Removed unused LOC

czhi-bin committed:
Merge pull request #2 from czhi-bin/branch-A-Assertions

Add Assertions in Duke and MainWindow

czhi-bin committed:
Merge branch 'master' into branch-A-CodeQuality

czhi-bin committed:
Merge pull request #3 from czhi-bin/branch-A-CodeQuality

Improve Code Quality in Parser

czhi-bin committed:
Merge branch 'master' into branch-A-Streams

czhi-bin committed:
Merge pull request #4 from czhi-bin/branch-A-Streams

Add usage of Stream in TaskList

czhi-bin committed:
Create gradle.yml

czhi-bin committed:
Fix checkstyle issue in test module

czhi-bin committed:
Add sort function

czhi-bin committed:
Merge branch 'branch-BCD-Extension'

czhi-bin committed:
Fix checkstyle issues.

czhi-bin committed:
Improve sort function to add reverse option.

Sort currently only have a default sorting sequence which is from
sorting task according to earlier date and time to later, then by
lexicographical order.

Some user might not want to go down the list just to see the latest
task and hence a reverse sort is needed such that latest task appear at
the top of the list.

Added functionality to check if the user input have "reverse".

User can choose to use sort without the reverse flag to sort the list
using default sorting sequence, and will not have to type "default".

czhi-bin committed:
Add Help function

Currently, there is no help function.

Some user might not fully understand how to use duke and a help message
which tells the user how to use duke and the commands would help them to
get used to the software faster.

Added help function.

The help command will show a help message showing all the command and
its required arguments and what are those arguments.

czhi-bin committed:
Fix JavaDoc and variable naming issues

czhi-bin committed:
Improve GUI

czhi-bin committed:
Change colour

czhi-bin committed:
Fix minor bug

czhi-bin committed:
Improve GUI

czhi-bin committed:
Adjust button size

czhi-bin committed:
Rename Duke to AngryBot

czhi-bin committed:
Fix minor bugs and update README.md

czhi-bin committed:
Update README.md

czhi-bin committed:
Set theme jekyll-theme-midnight

czhi-bin committed:
Fix minor bugs with sorting

czhi-bin committed:
Add new NoTaskDescriptionException

czhi-bin committed:
Merge branch 'master' of https://github.com/czhi-bin/ip

czhi-bin committed:
Change the jar name in build.gradle

czhi-bin committed:
Fix checkstyle issues

czhi-bin committed:
Change welcome message
